gax/status.h

```diff
@@ -0,0 +1,92 @@
+// Copyright 2019 Google Inc.  All rights reserved
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_GAX_STATUS_H_
+#define GOOGLE_GAX_STATUS_H_
+
+#include <ostream>
+#include <string>
+
+namespace google {
+namespace gax {
+
+/**
+ * The underlying StatusCode is intended to closely resemble
+ * grpc::StatusCode, the semantics of which are documented in:
+ *    https://grpc.io/grpc/cpp/classgrpc_1_1_status.html
+ */
+enum class StatusCode {
+  /// Not an error; returned on success.
+  kOk = 0,
+
+  kCancelled = 1,
+  kUnknown = 2,
+  kInvalidArgument = 3,
+  kDeadlineExceeded = 4,
+  kNotFound = 5,
+  kAlreadyExists = 6,
+  kPermissionDenied = 7,
+  kResourceExhausted = 8,
+  kFailedPrecondition = 9,
+  kAborted = 10,
+  kOutOfRange = 11,
+  kUnimplemented = 12,
+  kInternal = 13,
+  kUnavailable = 14,
+  kDataLoss = 15,
+  kUnauthenticated = 16,
+};
+
+/**
+ * Reports error code and details from a remote request.
+ *
+ * This class is modeled after `grpc::Status`.
+ * It contains the status code and error message(if applicable) from an RPC.
+ */
+class Status {
+ public:
+  Status() : code_(StatusCode::kOk) {}
+  explicit Status(StatusCode code, std::string msg) : code_(code),
+                                                      msg_(std::move(msg)) {}
+
+  inline bool IsOk() const { return code_ == StatusCode::kOk; }
+  inline bool IsTransientFailure() const {
```

Comment:
I would consider removing this and `IsPermanentFailure`. I'm not sure that they are very useful attached to the Status object, or whether any future absl::Status would have a similar function. For now let's leave them in to match google-cloud-cpp status, but worth thinking about whether we need them.

Reply:
What I'm imagining instead of having these three is instead keeping just `IsOk` and having transient/permanent detection handled by a separate idempotency policy that looks at the code and potentially keeps additional state or configuration. Is that what you were thinking of, or something else entirely?

Reply:
That is what I was thinking. My main reason for not suggesting we do that immediately is because google-cloud-cpp might want to use them, so we can keep them for now and consider removing after consulting with them.